usng4j-impl/src/main/java/org/codice/usng4j/impl/CoordinateSystemTranslatorImpl.java

```diff
@@ -552,24 +590,85 @@ BoundingBox toBoundingBox(final UtmCoordinate utmCoordinate, final Integer accur
     return result;
   }
 
-  /** {@inheritDoc} */
   @Override
-  public DecimalDegreesCoordinate toLatLon(UtmCoordinate utmCoordinate) {
-    if (utmCoordinate.getNSIndicator() == NSIndicator.SOUTH) {
-      UtmCoordinate newUtm =
-          new UtmCoordinateImpl(
-              utmCoordinate.getZoneNumber(),
-              utmCoordinate.getEasting(),
-              utmCoordinate.getNorthing() - NORTHING_OFFSET);
-      return toLatLonNsNormalized(newUtm);
+  public DecimalDegreesCoordinate toLatLon(final UtmUpsCoordinate utmUpsCoordinate) {
+    if (utmUpsCoordinate.isUTM()) {
+      return toLatLon((UtmCoordinate) utmUpsCoordinate);
+    } else {
+      return toLatLon((UpsCoordinate) utmUpsCoordinate);
+    }
+  }
+
+  @Override
+  public DecimalDegreesCoordinate toLatLon(final UpsCoordinate upsCoordinate) {
+    final double northing = upsCoordinate.getNorthing() - FALSE_UPS_NORTHING;
+    final double easting = upsCoordinate.getEasting() - FALSE_UPS_EASTING;
+
+    final boolean isNorth = upsCoordinate.getLatitudeBand() >= 'Y';
+
+    final double lat;
+    if (northing == 0.0 && easting == 0.0) {
+      lat = isNorth ? 90.0 : -90.0;
     } else {
-      return toLatLonNsNormalized(utmCoordinate);
+      final double rho = Math.hypot(easting, northing);
+      final double t = rho != 0.0 ? rho / RHO_ADJUSTER_VALUE : Math.pow(EPSILON, 2.0);
```

Comment:
✏️ I don't personally find ternaries with nested math processing especially easy to read. It is only a matter of preference, but when the get longer like this, I would much prefer they be expanded to if/else blocks.

Reply:
This was ported as is from GeographicLib, first into USNG.js and then here; we didn't want to break the actual calculations as we don't know all the math for it.